h2o-clustering/src/main/java/water/clustering/api/GracefulAsyncRunner.java

```diff
@@ -0,0 +1,40 @@
+package water.clustering.api;
+
+import fi.iki.elonen.NanoHTTPD;
+import org.apache.log4j.Logger;
+
+
+/**
+ * The REST API is only needed until a valid flatfile is received. Afterwards, the flatfile is handed over to
+ * H2O via embedded config. Therefore, as soon as the embedded config is constructed, the REST API is shut down
+ * and the objects destroyed. However, there might still be a thread running - sending HTTP response to the assist which
+ * sent the flatfile. The {@link fi.iki.elonen.NanoHTTPD.DefaultAsyncRunner} does not wait for existing connections
+ * to be properly terminated and shuts them down.
+ * 
+ * This implementation overrides that behavior and waits until all connections are closed before shutdown.
+ */
+public class GracefulAsyncRunner extends NanoHTTPD.DefaultAsyncRunner {
+    private static final Logger LOG = Logger.getLogger(GracefulAsyncRunner.class);
+
+    @Override
+    public void closeAll() {
+        while (getRunning().size() > 0) {
+            synchronized (this) {
+                try {
+                    this.wait();
```

Comment:
should there be a timeout in the wait() call to make sure this does not hang indefinitelly?

Reply:
I'd say it is not required, as by design the rest of the code just replies and no matter if the reply fails or succeds, the asyncRunner frees the resources and this part is unblocked. To make this futureproof, I've introduced a one minute timeout, which is PLENTY of space.

Reply:
I think you want a timeout here and a short one, because between calling `getRunning().size()` and this.wait() the state can change and you could be waiting forever

Reply:
True. Will make it one second - the while cycle is the guard..

Reply:
Done.